### PR TITLE
Mentors

### PR DESCRIPTION
**Pull request by ike709**
:cl: ike709
add: Mentors are a thing now. If you have a gameplay question, use mentorhelp. Don't bug admins about it. However, there probably aren't any mentors yet when this first gets merged.
/:cl:

Since I'm so terrible at bugfixing, I figured I'd compensate by creating more bugs.

Insert usual disclaimer about me not having any idea what I'm doing.

Note that this touches both the config and the database.

I've done some very basic testing. More testing is definitely required, but so far so good. I am also unable to test database stuff.

Also there's some new weird shit with how reading the config works and I'm not entirely sure I did that correctly.

### CODE REVIEW

**Comment by SmokingKilz** [comment]
Do it.

**Comment by AffectedArc07** [comment]
If this is merged how are applications gonna be handled?

**Comment by Ralta** [comment]
@AffectedArc07 DENIED.

On a serious note, is this going to be linked to the ticket system currently in place for admins at all? And if not, can it be?

File: code/modules/admin/verbs/adminhelp.dm [hunk 1]
```diff
@@ -501,6 +501,13 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
 			current_ticket.AddInteraction("[key_name_admin(usr)] opened a new ticket.")
 			current_ticket.Close()
 
+	//Extremely simple system of suggesting mentorhelp instead of adminhelp
```
Comment (Birdtalon) [inline code/modules/admin/verbs/adminhelp.dm]: This could potentially get very annoying.

File: code/modules/client/verbs/ooc.dm [hunk 2]
```diff
@@ -69,7 +69,10 @@
 				else
 					to_chat(C, "<font color='[GLOB.normal_ooc_colour]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[holder.fakekey ? holder.fakekey : key]:</EM> <span class='message'>[msg]</span></span></font>")
 			else if(!(key in C.prefs.ignoring))
-				to_chat(C, "<font color='[GLOB.normal_ooc_colour]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[keyname]:</EM> <span class='message'>[msg]</span></span></font>")
+				var/ooc_color = GLOB.normal_ooc_colour
+				if(check_mentor())
+					ooc_color = "#FF3E96"
```
Comment (Birdtalon) [inline code/modules/client/verbs/ooc.dm]: Maybe make this something that can be changed easily like the admin OOC colour?

File: code/modules/mentor/verbs/mentorhelp.dm [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	set name = "Mentorhelp"
+
+	//clean the input msg
+	if(!msg)
+		return
+	msg = sanitize(copytext(msg,1,MAX_MESSAGE_LEN))
+	if(!msg || !mob)
+		return
+	if(src.prefs.muted & MUTE_MENTORHELP)
+		to_chat(src, "<font color='red'>You are unable to use mentorhelp (muted).</font>")
```
Comment (Birdtalon) [inline code/modules/mentor/verbs/mentorhelp.dm]: Would `<span>` be better?

File: code/modules/mentor/verbs/mentorhelp.dm [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		return
+	var/show_char = CONFIG_GET(flag/mentors_mobname_only)
+	var/mentor_msg = "<span class='mentornotice'><b><font color='#3280ff'>MENTORHELP:</b> <b>[key_name_mentor(src, 1, 0, 1, show_char)]</b>:</font> [msg]</span>"
+	log_mentor("MENTORHELP: [key_name_mentor(src, 0, 0, 0, 0)]: [msg]")
+
+	for(var/client/X in GLOB.mentors)
+		X << 'sound/items/bikehorn.ogg'
```
Comment (Birdtalon) [inline code/modules/mentor/verbs/mentorhelp.dm]: Maybe we should add a sound specific to Mentorhelp?

**Comment by AffectedArc07** [comment]
Linking to the ticket system will probably be a huge pain, and not really that necessary, so I wouldn’t do it

File: code/modules/mentor/verbs/mentorhelp.dm (hunk 4 above)
Comment (Xantholne) [reply code/modules/mentor/verbs/mentorhelp.dm]: We can change admin bwoink to _adminhelp and give mentors adminhelp in the .ogg sound files
/s

**Comment by FlattestGuitar** [comment]
I don't really have the time to look into it properly, but there's a load of stuff that just works off of `client.holder` existing, which is the `/datum/admins` object, which you seem to set for mentors. Things like seeing dchat at all times and being able to view MC info that mentors definitely don't need falls under this.

While I'm not sure which things exactly are covered by that, doing a search for `.holder` or similar in the codebase should show most instances. I think it's best to set the rules on what mentors can and cannot see before we start with this mess.

Don't merge until we're sure that's taken care of, @AndrewMontagne, @SamHPurp

**Comment by ike709** [comment]
I'll get back to this after the maintainers have reviewed/merged/closed some of my other PRs.

**Comment by AndrewMontagne** [comment]
Ignore the database changelog comment. We've already fucked up and not kept it up to date. 🤔

**Comment by ike709** [comment]
@AndrewMontagne I'm pretty sure flat is totally wrong about that client.holder stuff because when I joined a local server with only the mentor rank, I only had access to regular player stuff and the mentor panel. No MC info or anything like that.

File: code/modules/mentor/verbs/mentorhelp.dm (hunk 4 above)
Comment (AffectedArc07) [reply code/modules/mentor/verbs/mentorhelp.dm]: I think something other than the bikehorn would be better. Maybe something like the announcement sound?

**Comment by ike709** [comment]
I was right, flat was totally wrong.

**Comment by ike709** [comment]
All went well when Andrew and I tested it.

File: code/modules/mentor/verbs/mentorhelp.dm (hunk 3 above)
Comment (FlattestGuitar) [reply code/modules/mentor/verbs/mentorhelp.dm]: Yeah, make that into `<span class='danger'>`

File: code/modules/mentor/verbs/mentorpm.dm [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	else if(istype(whom,/client))
+		C = whom
+	if(!C)
+		if(holder)
+			to_chat(src, "<font color='red'>Error: Mentor-PM: Client not found.</font>")
+		else
+			mentorhelp(msg)	//Mentor we are replying to left. Mentorhelp instead
+		return
+
+	//get message text, limit it's length.and clean/escape html
```
Comment (FlattestGuitar) [inline code/modules/mentor/verbs/mentorpm.dm]: its*

File: code/modules/mentor/verbs/mentorhelp.dm [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		to_chat(X, mentor_msg)
+
+	for(var/client/A in GLOB.admins)
+		A << 'sound/items/bikehorn.ogg'
+		to_chat(A, mentor_msg)
+
+	to_chat(src, "<span class='mentornotice'><font color='purple'>PM to-<b>Mentors</b>: [msg]</font></span>")
```
Comment (FlattestGuitar) [inline code/modules/mentor/verbs/mentorhelp.dm]: You use the span class `mentornotice` twice, but it's not defined. Make sure it's in both `interface/stylesheet.dm` and `code/modules/goonchat/browserassets/css/browserOutput.css`

**Comment by FlattestGuitar** [comment]
Also, in general, please don't use `<font color=X>`. Replace all instances with span classes. Make new ones if you must, but use preexisting ones if you can.

**Comment by FlattestGuitar** [comment]
Otherwise LGTM. Once those issues are fixed, feel free to smash that merge button and subscribe. @AndrewMontagne

**Comment by AffectedArc07** [comment]
@ike709 travis machine broke

**Comment by ike709** [comment]
not my fault

**Comment by AffectedArc07** [comment]
Where do I sign up

**Comment by ike709** [comment]
Not here. Go away.

**Pull request merged**